### PR TITLE
Recomputable

### PR DESCRIPTION
**Pull request by strake**
Given a decidable relation, we can rebuild a relevant proof given an irrelevant one. Please feel free to propose a better name if you have one! I wasn't sure what to call it.

### CODE REVIEW

**Comment by gallais** [comment]
Note that this type alias could readily be used in:

* `Relation.Nullary`'s [`recompute`](https://agda.github.io/agda-stdlib/Relation.Nullary.html#782)

* `Data.Rational.Properties`' [`recomputeCP`](https://agda.github.io/agda-stdlib/Data.Rational.Properties.html#1225)

I don't know what a good name for this principle would be though.

**Comment by strake** [comment]
@gallais We could use it in `Data.Rational.Properties` but i'm not sure how much we gain for a private lemma — `Irrelevant′ (λ n d → Coprime n (suc d))` is not much shorter than `∀ {n d} → .(Coprime n (suc d)) → Coprime n (suc d)`.

File: src/Relation/Binary/Consequences.agda [hunk 1]
```diff
@@ -164,3 +164,9 @@ module _ {P : REL A B ℓ₁} {Q : REL B A ℓ₂} where
 
   flip-Connex : Connex P Q → Connex Q P
   flip-Connex f x y = Sum.swap (f y x)
+
+module _ {a b r} {A : Set a} {B : Set b} {R : REL A B r} where
+     dec⟶irrel′ : Decidable R → Irrelevant′ R
```
Comment (gallais) [inline src/Relation/Binary/Consequences.agda]: You can drop `{a b} {A : Set a} {B : Set b}` thanks to [the variables](https://github.com/agda/agda-stdlib/blob/5741c80877e7d3ffcee9a842804ccb794719d3d3/src/Relation/Binary/Consequences.agda#L21).

File: src/Relation/Binary/Consequences.agda [hunk 2]
```diff
@@ -164,3 +164,9 @@ module _ {P : REL A B ℓ₁} {Q : REL B A ℓ₂} where
...
+     dec⟶irrel′ dec {a} {b} _ with dec a b
+     dec⟶irrel′ _ _ | yes p = p
+     dec⟶irrel′ _ r | no ¬p = ⊥-elim (¬p r)
```
Comment (gallais) [inline src/Relation/Binary/Consequences.agda]: `dec⟶irrel′` could be implemented using [`recompute`](https://agda.github.io/agda-stdlib/Relation.Nullary.html#782), couldn't it?

**Comment by strake** [comment]
@gallais All done

**Comment by MatthewDaggitt** [comment]
Hmm for starters I agree with @strake that the current name isn't great. Perhaps something like `Recomputable`? 

My second question is what's the use case for this? Are there instances of common non-decidable relations for which this holds? If not, and it's only useful in the two recomputable lemmas then perhaps it doesn't deserve it's own type?

**Comment by strake** [comment]
@MatthewDaggitt The obvious example to me is pointwise equality of functions given recomputable equality of the codomain:
```agda
module _ {a b} {A : Set a} {B : A → Set b} where
    _≃_ : Rel ((a : A) → B a) (a ⊔ b)
    f ≃ g = ∀ {a} → f a ≡ g a

    ≃-irrel′ : (∀ {a} → Irrelevant′ (_≡_ {_} {B a})) → Irrelevant′ _≃_
    ≃-irrel′ d x = d x

    ≃-dec : (∀ {a} → Decidable (_≡_ {_} {B a})) → Decidable _≃_
    ≃-dec d f g = {!!} -- I not believe it's provable.
```

**Comment by strake** [comment]
And "`Recomputable`" is fine for me.

**Comment by strake** [comment]
@MatthewDaggitt Renamed as `Recomputable`.

File: CHANGELOG.md [hunk 3]
```diff
@@ -308,10 +308,10 @@ Other minor additions
...
+* Defined a notion of recomputability for binary relations:
   ```agda
-  Irrelevant′ : REL A B ℓ → Set _
-  Irrelevant′ _~_ = ∀ {x y} → .(x ~ y) → x ~ y
+  Recomputable : REL A B ℓ → Set _
+  Recomputable _~_ = ∀ {x y} → .(x ~ y) → x ~ y
 
-  dec⟶irrel′ : Decidable R → Irrelevant′ R
+  dec⟶irrel′ : Decidable R → Recomputable R
   ```
```
Comment (gallais) [inline CHANGELOG.md]: the name should be changed in the CHANGELOG too

Comment (strake) [reply CHANGELOG.md]: done

**Comment by MatthewDaggitt** [comment]
> Edit: This includes certain representations of ℝ, for example: ℕ → Bool (binary expansion, Stern-Brocot, whatever you like).

Thanks, that's a great example! Would you be able to add the Unary version as well for completeness? Also as it's a bit of an uncommon name I think the full name should probably be used, i.e. `dec⟶recomputable` rather than `dec⟶recomput`.

Once that's done, I'll merge it in.

**Comment by strake** [comment]
@MatthewDaggitt All done ☺

**Pull request merged**